src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/ActionExplorationDriver.cs

```diff
@@ -0,0 +1,207 @@
+     */
+    public class ActionExplorationDriver : MonoBehaviour
+    {
+        private const int PRIOR_ACTION_LIMIT = 5;
+
+        // normally the first thing we do is retry the previous actions
+        // these are sorted newest to oldest
+        private readonly List<IBotActionData> PreviouslyCompletedActions = new(PRIOR_ACTION_LIMIT);
+
+        private int _previousActionsNextIndex = 0;
+        private int _previousActionsStartIndex = 0;
+
+        public ExplorationState ExplorationState { get; private set; } = ExplorationState.STOPPED;
+
+        public void ReportPreviouslyCompletedAction(IBotActionData action)
+        {
+            // limit to PRIOR_ACTION_LIMIT prior actions
+            while (PreviouslyCompletedActions.Count >= PRIOR_ACTION_LIMIT)
+            {
+                // remove at front
+                PreviouslyCompletedActions.RemoveAt(0);
+            }
+
+            if (!PreviouslyCompletedActions.Contains(action))
+            {
+                var extraLog = "";
+                if (action is KeyMomentMouseActionData keyMomentMouseActionData)
+                {
+                    extraLog += " with first object path: " + keyMomentMouseActionData.mouseActions[1].clickedObjectNormalizedPaths[0];
+                }
+                RGDebug.LogInfo($"ActionExplorationDriver - Adding previously completed action of Type: {action.GetType().Name}" + extraLog);
+                // insert at end
+                PreviouslyCompletedActions.Add(action);
+                _previousActionsNextIndex = PreviouslyCompletedActions.Count-1;
+                _previousActionsStartIndex = PreviouslyCompletedActions.Count-1;
+            }
+        }
+
+        public void StartExploring()
+        {
+            switch (ExplorationState)
+            {
+                case ExplorationState.STOPPED:
+                    _previousActionsNextIndex = PreviouslyCompletedActions.Count-1;
+                    _previousActionsStartIndex = PreviouslyCompletedActions.Count-1;
+                    RGDebug.LogInfo("ActionExplorationDriver - Starting Exploratory Actions");
+                    ExplorationState = ExplorationState.EXPLORING;
+                    break;
+                case ExplorationState.PAUSED:
+                    RGDebug.LogInfo("ActionExplorationDriver - Resuming Exploratory Actions");
+                    ExplorationState = ExplorationState.EXPLORING;
+                    break;
+
+            }
+
+        }
+
+        private IBotActionData _inProgressAction = null;
+
+        /**
+         * Chooses the next exploratory action to perform.
+         *
+         * The current implementation chooses one of the previous N successful IKeyMomentExploration actions to try on each update pass.
+         */
+        public void PerformExploratoryAction(int segmentNumber, Dictionary<long, ObjectStatus> currentTransforms, Dictionary<long, ObjectStatus> currentEntities, out string error)
```

Comment:
This is where exploratory actions are driven from.  We will hook in other exploratory action types to this class as we create them